ambari-server/src/main/java/org/apache/ambari/server/controller/internal/ComponentResourceProvider.java

```diff
@@ -413,21 +414,16 @@ private ServiceComponentRequest getRequest(Map<String, Object> properties) {
 
       debug("Received a createComponent request: {}", request);
 
-      if (!componentNames.containsKey(request.getClusterName())) {
-        componentNames.put(request.getClusterName(), new HashMap<>());
-      }
-
-      Map<String, Set<String>> serviceComponents = componentNames.get(request.getClusterName());
-      if (!serviceComponents.containsKey(request.getServiceName())) {
-        serviceComponents.put(request.getServiceName(), new HashSet<String>());
-      }
+      Set<String> componentID = ImmutableSet.of(request.getServiceGroupName(), request.getServiceName(), request.getComponentName());
+      boolean added = componentNames
```

Comment:
Shouldn't this be a list? (ordering matters)

Reply:
Nice catch.